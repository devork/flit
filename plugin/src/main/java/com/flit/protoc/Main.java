package com.flit.protoc;

import com.google.protobuf.compiler.PluginProtos;

/**
 * Main entry point to the plugin which handles the parsing of the input from the main protoc process
 * and hands off to a {@link Plugin} instance.
 */
public class Main {

  public static void main(String[] args) throws Exception {
    Plugin plugin = new Plugin(PluginProtos.CodeGeneratorRequest.newBuilder().mergeFrom(System.in).build());
    plugin.process().writeTo(System.out);
  }

}
