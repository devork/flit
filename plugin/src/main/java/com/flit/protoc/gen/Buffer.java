package com.flit.protoc.gen;

/**
 * Holds a generation buffer used to write basic strings to a wrapper string builder
 */
public class Buffer {

  private StringBuilder buffer;
  private Indent indent;

  public Buffer() {
    this.indent = new Indent();
    this.buffer = new StringBuilder();
  }

  // write
  public Buffer w(String s) {
    buffer.append(s);
    return this;
  }

  // write (multi)
  public Buffer w(String... ss) {

    for (String s : ss) {
      buffer.append(s);
    }
    return this;
  }

  // indent write
  public Buffer iw(String s) {
    buffer.append(indent).append(s);
    return this;
  }

  // indent write (multi)
  public Buffer iw(String... ss) {
    buffer.append(indent);

    for (String s : ss) {
      buffer.append(s);
    }
    return this;
  }

  // write newline
  public Buffer wn(String s) {
    buffer.append(s).append("\n");
    return this;
  }

  // write newline (multi)
  public Buffer wn(String... ss) {

    for (String s : ss) {
      buffer.append(s);
    }
    buffer.append("\n");
    return this;
  }

  // indent write newline
  public Buffer iwn(String s) {
    buffer.append(indent).append(s).append("\n");
    return this;
  }

  // indent write newline (multi)
  public Buffer iwn(String... ss) {
    buffer.append(indent);

    for (String s : ss) {
      buffer.append(s);
    }
    buffer.append("\n");
    return this;
  }

  // newline
  public Buffer n() {
    buffer.append("\n");
    return this;
  }

  // increase the indentation
  public Buffer inc() {
    indent.inc();
    return this;
  }

  // decrease the indentation
  public Buffer dec() {
    indent.dec();
    return this;
  }

  public String toString() {
    return this.buffer.toString();
  }

}
