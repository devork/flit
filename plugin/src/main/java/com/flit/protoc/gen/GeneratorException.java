package com.flit.protoc.gen;

public class GeneratorException extends RuntimeException {
  public GeneratorException(String message) {
    super(message);
  }
}
