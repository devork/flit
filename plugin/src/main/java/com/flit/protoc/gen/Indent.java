package com.flit.protoc.gen;

public class Indent {
  private final String indent;
  private String current = "";

  public Indent() {
    this.indent = "    ";
  }

  public Indent(String indent) {
    this.indent = indent;
  }

  public String inc() {
    current += indent;
    return current;
  }

  public String dec() {
    if (current.isEmpty()) {
      return current;
    }

    current = current.substring(indent.length());
    return current;
  }

  public String toString() {
    return current;
  }
}
