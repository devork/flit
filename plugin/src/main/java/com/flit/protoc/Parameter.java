package com.flit.protoc;

import com.flit.protoc.gen.GeneratorException;
import lombok.Getter;
import lombok.ToString;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Getter @ToString public class Parameter {

  public static final String PARAM_TARGET = "target";
  public static final String PARAM_CLIENT = "client";
  public static final String PARAM_TYPE = "type";
  public static final String PARAM_CONTEXT = "context";

  private String key;
  private String value;

  public Parameter(String[] strings) {
    this.key = strings[0];
    this.value = strings[1];
  }

  public static Map<String, Parameter> of(String value) {
    if (value == null || (value = value.trim()).isEmpty()) {
      throw new GeneratorException("Empty value passed to parameter builder");
    }

    return Arrays.stream(value.split(","))
      .map(p -> p.split("=", 2))
      .filter(s -> s.length == 2 && !s[0].trim().isEmpty() && !s[1].trim().isEmpty())
      .map(Parameter::new)
      .collect(Collectors.toMap(Parameter::getKey, Function.identity()));
  }
}
