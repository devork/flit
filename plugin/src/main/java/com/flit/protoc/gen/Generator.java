package com.flit.protoc.gen;

import com.flit.protoc.Parameter;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse;

import java.util.List;
import java.util.Map;

/**
 * Defines the handler for the generator step
 */
public interface Generator {

  /**
   * This is the main entry point to code generation: the implementation is decided in the {@link com.flit.protoc.Plugin}
   * class from the given parameters.
   *
   * @param request   The inbound protoc request
   * @param params    The plugin parameters
   *
   * @return The list of files to be added to the output.
   */
  List<CodeGeneratorResponse.File> generate(CodeGeneratorRequest request, Map<String, Parameter> params);
}
