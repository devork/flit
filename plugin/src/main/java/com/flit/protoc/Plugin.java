package com.flit.protoc;

import com.flit.protoc.gen.Generator;
import com.flit.protoc.gen.GeneratorException;
import com.flit.protoc.gen.server.spring.SpringGenerator;
import com.flit.protoc.gen.server.undertow.UndertowGenerator;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse;

import java.util.Map;

import static com.flit.protoc.Parameter.PARAM_TARGET;
import static com.flit.protoc.Parameter.PARAM_TYPE;

public class Plugin {

  private final CodeGeneratorRequest request;

  public Plugin(CodeGeneratorRequest request) {
    this.request = request;
  }

  public CodeGeneratorResponse process() {
    if (!request.hasParameter()) {
      return CodeGeneratorResponse.newBuilder().setError("Usage: --flit_out=target=server,type=[spring|undertow]:<PATH>").build();
    }

    Map<String, Parameter> params = Parameter.of(request.getParameter());
    try {
      CodeGeneratorResponse.Builder builder = CodeGeneratorResponse.newBuilder();
      resolveGenerator(params).generate(request, params).forEach(builder::addFile);
      return builder.build();
    } catch (GeneratorException e) {
      return CodeGeneratorResponse.newBuilder().setError(e.getMessage()).build();
    }
  }

  private Generator resolveGenerator(Map<String, Parameter> params) {
    if (!params.containsKey(PARAM_TARGET)) {
      throw new GeneratorException("No argument specified for target");
    }
    if (!params.containsKey(PARAM_TYPE)) {
      throw new GeneratorException("No argument specified for type");
    }
    switch (params.get(PARAM_TARGET).getValue()) {
      case "server":
        switch (params.get(PARAM_TYPE).getValue()) {
          case "boot":
          case "spring":
            return new SpringGenerator();
          case "undertow":
            return new UndertowGenerator();
          default:
            throw new GeneratorException("Unknown server type: " + params.get(PARAM_TYPE).getValue());
        }
      default:
        throw new GeneratorException("Unknown target type: " + params.get(PARAM_TARGET).getValue());
    }
  }
}
