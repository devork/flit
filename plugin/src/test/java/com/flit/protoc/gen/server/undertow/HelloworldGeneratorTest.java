package com.flit.protoc.gen.server.undertow;

import com.flit.protoc.Plugin;
import com.flit.protoc.gen.BaseGeneratorTest;
import com.google.protobuf.compiler.PluginProtos;
import org.approvaltests.Approvals;
import org.junit.Test;

import static java.util.stream.Collectors.toList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class HelloworldGeneratorTest extends BaseGeneratorTest {

  @Test public void test_Generate() throws Exception {
    PluginProtos.CodeGeneratorRequest request = loadJson("helloworld.undertow.json");

    Plugin plugin = new Plugin(request);
    PluginProtos.CodeGeneratorResponse response = plugin.process();

    assertNotNull(response);
    assertEquals(2, response.getFileCount());
    assertEquals(response.getFile(0).getName(), "com/example/helloworld/RpcHelloWorld.java");
    assertEquals(response.getFile(1).getName(), "com/example/helloworld/RpcHelloWorldHandler.java");

    Approvals.verifyAll("", response.getFileList().stream().map(f -> f.getContent()).collect(toList()));
    response.getFileList().forEach(f -> assertParses(f));
  }

}
