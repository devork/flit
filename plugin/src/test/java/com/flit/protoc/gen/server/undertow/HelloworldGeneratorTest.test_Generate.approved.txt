[0] = package com.example.helloworld;

public interface RpcHelloWorld {
  Helloworld.HelloResp handleHello(Helloworld.HelloReq in);

  Helloworld.HelloResp handleHelloAgain(Helloworld.HelloReq in);
}

[1] = package com.example.helloworld;

import com.flit.runtime.ErrorCode;
import com.flit.runtime.FlitException;
import com.flit.runtime.undertow.ErrorWriter;
import com.flit.runtime.undertow.FlitHandler;
import com.google.protobuf.util.JsonFormat;
import io.undertow.server.HttpHandler;
import io.undertow.server.HttpServerExchange;
import io.undertow.util.Headers;
import java.io.InputStreamReader;
import java.lang.Exception;
import java.lang.Override;
import java.lang.String;
import java.nio.charset.StandardCharsets;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RpcHelloWorldHandler implements HttpHandler {
  private static final Logger LOGGER = LoggerFactory.getLogger(RpcHelloWorldHandler.class);

  public static final String ROUTE = "/twirp/com.example.helloworld.HelloWorld";

  private final RpcHelloWorld service;

  private final ErrorWriter errorWriter;

  public RpcHelloWorldHandler(RpcHelloWorld service) {
    this.service = service;
    this.errorWriter = new ErrorWriter();
  }

  @Override
  public void handleRequest(HttpServerExchange exchange) throws Exception {
    if (exchange.isInIoThread()) {
      exchange.dispatch(this);
      return;
    }
    exchange.startBlocking();
    String method = exchange.getAttachment(FlitHandler.KEY_METHOD);
    try {
      switch (method) {
        case "Hello": handleHello(exchange); break;
        case "HelloAgain": handleHelloAgain(exchange); break;
        default: throw FlitException.builder().withErrorCode(ErrorCode.BAD_ROUTE).withMessage("No such route").build();
      }
    } catch (FlitException e) {
      errorWriter.write(e, exchange);
    } catch (Exception e) {
      LOGGER.error("Exception caught at handler: {}", e.getMessage(), e);
      errorWriter.write(e, exchange);
    }
  }

  private void handleHello(HttpServerExchange exchange) throws Exception {
    boolean json = false;
    final Helloworld.HelloReq data;
    final String contentType = exchange.getRequestHeaders().get(Headers.CONTENT_TYPE).getFirst();
    if (contentType.equals("application/protobuf")) {
      data = Helloworld.HelloReq.parseFrom(exchange.getInputStream());
    } else if (contentType.startsWith("application/json")) {
      json = true;
      Helloworld.HelloReq.Builder builder = Helloworld.HelloReq.newBuilder();
      JsonFormat.parser().merge(new InputStreamReader(exchange.getInputStream(), StandardCharsets.UTF_8), builder);
      data = builder.build();
    } else {
      exchange.setStatusCode(415);
      return;
    }
    Helloworld.HelloResp response = service.handleHello(data);
    exchange.setStatusCode(200);
    if (json) {
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/json;charset=UTF-8");
      exchange.getResponseSender().send(JsonFormat.printer().omittingInsignificantWhitespace().print(response));
    } else {
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/protobuf");
      response.writeTo(exchange.getOutputStream());
    }
  }

  private void handleHelloAgain(HttpServerExchange exchange) throws Exception {
    boolean json = false;
    final Helloworld.HelloReq data;
    final String contentType = exchange.getRequestHeaders().get(Headers.CONTENT_TYPE).getFirst();
    if (contentType.equals("application/protobuf")) {
      data = Helloworld.HelloReq.parseFrom(exchange.getInputStream());
    } else if (contentType.startsWith("application/json")) {
      json = true;
      Helloworld.HelloReq.Builder builder = Helloworld.HelloReq.newBuilder();
      JsonFormat.parser().merge(new InputStreamReader(exchange.getInputStream(), StandardCharsets.UTF_8), builder);
      data = builder.build();
    } else {
      exchange.setStatusCode(415);
      return;
    }
    Helloworld.HelloResp response = service.handleHelloAgain(data);
    exchange.setStatusCode(200);
    if (json) {
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/json;charset=UTF-8");
      exchange.getResponseSender().send(JsonFormat.printer().omittingInsignificantWhitespace().print(response));
    } else {
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/protobuf");
      response.writeTo(exchange.getOutputStream());
    }
  }
}

